File: JetpackLiveDataJavaVersion/app/src/main/java/com/teletian/jetpacklivedatajavaversion/NetworkLiveData.java
package com.teletian.jetpacklivedatajavaversion;

import android.annotation.SuppressLint;
import android.content.BroadcastReceiver;
import android.content.Context;
import android.content.Intent;
import android.content.IntentFilter;
import android.net.ConnectivityManager;
import android.net.NetworkInfo;

import androidx.lifecycle.LiveData;

public class NetworkLiveData extends LiveData<NetworkInfo> {

    private Context context;
    private IntentFilter intentFilter = new IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION);
    private NetworkReceiver networkReceiver = new NetworkReceiver();

    @SuppressLint("StaticFieldLeak")
    private static NetworkLiveData instance;

    private NetworkLiveData(Context context) {
        this.context = context;
    }

    public static NetworkLiveData getInstance(Context context) {
        if (instance == null) {
            instance = new NetworkLiveData(context.getApplicationContext());
        }
        return instance;
    }

    @Override
    protected void onActive() {
        super.onActive();
        context.registerReceiver(networkReceiver, intentFilter);
    }

    @Override
    protected void onInactive() {
        super.onInactive();
        context.unregisterReceiver(networkReceiver);
    }

    static class NetworkReceiver extends BroadcastReceiver {

        @Override
        public void onReceive(Context context, Intent intent) {
            ConnectivityManager connectivityManager =
                    (ConnectivityManager) context.getSystemService(Context.CONNECTIVITY_SERVICE);
            NetworkInfo networkInfo = connectivityManager.getActiveNetworkInfo();
            NetworkLiveData.getInstance(context).setValue(networkInfo);
        }
    }
}

File: JetpackLiveData/app/src/main/java/com/teletian/jetpacklivedata/NetworkLiveData.kt
package com.teletian.jetpacklivedata

import android.content.BroadcastReceiver
import android.content.Context
import android.content.Intent
import android.content.IntentFilter
import android.net.ConnectivityManager
import android.net.NetworkInfo
import android.util.Log
import androidx.annotation.MainThread
import androidx.lifecycle.LiveData

class NetworkLiveData(private val context: Context?) : LiveData<NetworkInfo?>() {

    private var networkReceiver: NetworkReceiver
    private var intentFilter: IntentFilter

    init {
        networkReceiver = NetworkReceiver()
        intentFilter = IntentFilter(ConnectivityManager.CONNECTIVITY_ACTION)
    }

    companion object {
        private lateinit var instance: NetworkLiveData

        @MainThread
        fun getInstance(context: Context?): NetworkLiveData {
            instance = if (::instance.isInitialized) instance else NetworkLiveData(context)
            return instance
        }
    }

    override fun onActive() {
        super.onActive()
        Log.d("tianjf", "onActive")
        context!!.applicationContext.registerReceiver(networkReceiver, intentFilter)
    }

    override fun onInactive() {
        super.onInactive()
        Log.d("tianjf", "onInactive")
        context!!.applicationContext.unregisterReceiver(networkReceiver)
    }

    inner class NetworkReceiver : BroadcastReceiver() {
        override fun onReceive(context: Context?, intent: Intent?) {
            val connectivityManager =
                context?.getSystemService(Context.CONNECTIVITY_SERVICE) as? ConnectivityManager
            val networkInfo = connectivityManager?.activeNetworkInfo
            getInstance(context).value = networkInfo
        }

    }
}